File: 014/LongestCommonPrefix.scala
object LongestCommonPrefix {
    def longestCommonPrefix(strs: Array[String]): String = {
        @annotation.tailrec
        def loop(i: Int): Int = if (i == strs.head.length) i 
        else {
            val c = strs.head(i)
            if (strs.exists { str => 
                str.length == i || str(i) != c
            }) i 
            else loop(i + 1)
        }

        if (strs.isEmpty) ""
        else {
            val longestCommonPrefixLength = loop(0)
            strs.head.substring(0, longestCommonPrefixLength)
        }
    }

    def main(args: Array[String]) {
        var strs = Array("flower", "flow", "flight")
        println (longestCommonPrefix(strs))
    }
}

File: 014/longest-common-prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
    if (!strs.length)
        return "";

    function loop(i) {
        if (i == strs[0].length)
            return i;

        let c = strs[0].charAt(i);
        for (let str of strs) {
            if (str.length == i || str[i] != c)
                return i;
        }
        return loop(i+1);
    }

    let longestCommonPrefixLength = loop(0);
    return strs[0].substring(0, longestCommonPrefixLength);
};

let strs = ["flower", "flow", "flight"];
console.log(longestCommonPrefix(strs));
